File: fm.py
'''
    FM modulation
    author: Mahmoud
'''

import numpy as np
from scipy.io import wavfile
import librosa

from demod import fmdemod


def calc_kf(beta, m, sample_rate):
    '''carson rule:
    kf = β * B * 2π / mp
    '''
    max_freq = sample_rate/2

    mp = max(abs(m))

    return beta * max_freq * 2 * np.pi / mp


def modulate_fm(audio, sample_rate, beta):
    '''modulated(t) = Ac * cos(wc * t + kf * integration(audio(t)))'''
    # time samples
    time = np.arange(0, audio.size) / sample_rate

    ac = 1

    fc = 100 * 1000  # 100 kHz
    wc = 2 * np.pi * fc

    kf = calc_kf(beta, audio, sample_rate)

    return (ac * np.cos(wc * time + kf * np.cumsum(audio))), (kf * abs(audio.max()) / 2 * np.pi)


print('read audio file')
audio, sample_rate = librosa.load('sample.wav')


for beta, name in [(5, 'wide'), (.1, 'narrow')]:
    print(f'modeulate {name} band with β = {beta}')
    modulated, delta_f = modulate_fm(audio, sample_rate, beta)

    for snr in [0, 1, 10, 20]:
        out_file_name = f'out/fm_{name}_snr_{snr}.wav'
        print(out_file_name, end='', flush=True)

        # add noise
        noise = np.random.normal(0, 1/snr if snr != 0 else 1, len(audio))
        modulated_with_noise = modulated + noise

        # demodulate
        demodulated = fmdemod(modulated_with_noise,
                              sample_rate, delta_f, 100 * 1000)

        # out
        wavfile.write(out_file_name, sample_rate, demodulated)

        print(': done')

File: fm.m
%
% FM modulation
% author: Mahmoud Adas
%

pkg load communications;

function modulated = fmmod2(x, fc, fs, freqdev)
    t = (0:1/fs:((size(x,1)-1)/fs))';
    t = t(:,ones(1,size(x,2)));

    int_x = cumsum(x)/fs;
    modulated = cos(2*pi*fc*t + 2*pi*freqdev*int_x);
endfunction

function demodulated = fmdemod2(y, fc, fs, freqdev)
    t = (0:1/fs:((size(y,1)-1)/fs))';
    t = t(:,ones(1,size(y,2)));

    yq = hilbert(y).*exp(-j*2*pi*fc*t);
    demodulated = (1/(2*pi*freqdev))*[zeros(1,size(yq,2)); diff(unwrap(angle(yq)))*fs];
endfunction

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

disp('read audio file');
[audio, sample_rate] = audioread('sample.wav');

disp('stereo to mono');
audio = (audio(:, 1) + audio(:, 2)) / 2;

fc = 100 * 1000  % 100 kHz

disp('resample')
orig_sample_rate = sample_rate
sample_rate = fc*4
audio = resample(audio, sample_rate, orig_sample_rate);

betas.wide = 5;
betas.narrow = .1;
for [beta_value, name] = betas
    beta_value
    freqdev = beta_value * sample_rate

    disp('modulate')
    modulated = fmmod2(audio, fc, sample_rate, freqdev);

    for snr = [20, 10, 1, 0]
        out_file_name = ['out/fm_' name '_snr_'  num2str(snr) '.wav'];
        disp(out_file_name);

        disp('  add noise')
        modulated_with_noise = awgn(modulated, snr);

        disp('  demodulate')
        demodulated = fmdemod2(modulated_with_noise, fc, sample_rate, freqdev);

        disp('  resample back')
        demodulated = resample(demodulated, orig_sample_rate, sample_rate);

        disp('  write out')
        audiowrite(out_file_name, demodulated, orig_sample_rate);

        disp('  done');
    endfor
endfor
